specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,391 @@
+    - [`depositTransaction`](#deposittransaction)
+    - [`setGasPayingToken`](#setgaspayingtoken)
+  - [StandardBridge](#standardbridge)
+  - [CrossDomainMessenger](#crossdomainmessenger)
+  - [SystemConfig](#systemconfig)
+  - [L1Block](#l1block)
+- [User Flow](#user-flow)
+- [Upgrade](#upgrade)
+  - [L1Block Deployment](#l1block-deployment)
+  - [L2CrossDomainMessenger Deployment](#l2crossdomainmessenger-deployment)
+  - [L2StandardBridge Deployment](#l2standardbridge-deployment)
+  - [L1Block Proxy Update](#l1block-proxy-update)
+  - [L2CrossDomainMessenger Proxy Update](#l2crossdomainmessenger-proxy-update)
+  - [L2StandardBridge Proxy Update](#l2standardbridge-proxy-update)
+- [Selection of `ETHER_TOKEN_ADDRESS`](#selection-of-ether_token_address)
+- [Standard Config](#standard-config)
+- [Security Considerations](#security-considerations)
+  - [OptimismPortal Token Allowance](#optimismportal-token-allowance)
+  - [Decimal Scaling](#decimal-scaling)
+  - [Interoperability Support](#interoperability-support)
+  - [Wrapped Ether](#wrapped-ether)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Custom gas token is also known as "custom L2 native asset". It allows for an L1-native ERC20 token to collateralize
+and act as the gas token on L2. The native asset is used to buy gas which is used to pay for resources on the network
+and is represented by EVM opcodes such as `CALLVALUE` (`msg.value`).
+
+Both the L1 and L2 smart contract systems MUST be able to introspect on the address of the gas paying token to
+ensure the security of the system. The source of truth for the configuration of the address of the token is
+in the L1 `SystemConfig` smart contract.
+
+The terms "custom gas token", "gas paying token", "native asset" and "gas paying asset" can all be used interchangebly.
+Ether is not a token.
+
+## Constants
+
+| Constant          | Value                           | Description |
+| ----------------- | ------------------------------- | ----------- |
+| `ETHER_TOKEN_ADDRESS` |  `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` | Represents ether for gas paying asset |
+| `DEPOSITOR_ACCOUNT` | `address(0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001)` | Account with auth permissions on `L1Block` contract |
+| `GAS_PAYING_TOKEN_SLOT` | `bytes32(uint256(keccak256("opstack.gaspayingtoken")) - 1)` | Storage slot that contains the address and decimals of the gas paying token |
+
+## Properties of a Gas Paying Token
+
+The gas paying token MUST satisfy the standard [ERC20](https://eips.ethereum.org/EIPS/eip-20) interface
+and implementation. It MUST be a L1 contract.
+
+The gas paying token cannot:
+
+- Have a fee on transfer
+- Have rebasing logic
+- Have call hooks on transfer
+- Have more than 18 decimals
+- Have out of band methods for modifying balance or allowance
+
+It MUST be enforced on chain that the token has 18 or less decimals to guarantee no losses of precision when
```

Comment:
If we do not verify it on chain, then we can have loss of precision for user deposits and they will lose out on a small amount of value that will get stuck in the portal and not be minted on the L2. This is because of the way that the scaling logic works